aesd-char-driver/main.c

```diff
@@ -0,0 +1,228 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+#include "aesdchar.h"
+#include <linux/slab.h> 
+#include <linux/uio.h> 
+#include "aesd-circular-buffer.h"
+
+
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Shweta Prasad"); 
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{   struct aesd_dev *dev;
+    PDEBUG("open");
+    
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);   
+    filp->private_data = dev;
+    
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    struct aesd_dev *dev = filp->private_data;
+    struct aesd_buffer_entry *data;
+    size_t bytes_read;
+    size_t bytes_to_read = bytes_read; 
+    size_t i;
+
+    
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+
+    mutex_lock(&dev->lock);
+
+    // Retrieve the pointer to data from the circular buffer starting at *f_pos
+    data = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->buffer, *f_pos, &bytes_read);
+    if (!data) 
+    {
+        mutex_unlock(&dev->lock);
+        return 0; // No data available
+    }
+
+   
+    if (bytes_to_read > count) 
+    {
+        bytes_to_read = count; 
+    }
+
+
+    // Transfer data byte by byte to the user space buffer using put_user
+    for (i = 0; i < bytes_to_read; i++) 
+    {
+        if (put_user(data->buffptr[i], buf + i)) 
+        {
+            retval = -EFAULT;
+            mutex_unlock(&dev->lock);
+            return retval;
+        }
+    }
+
+    *f_pos += bytes_to_read; // Update the file position
+    retval = bytes_to_read;   // Return the number of bytes read
+
+    mutex_unlock(&dev->lock);
```

Comment:
Could move unlock up above the `*f_pos` line since lines 94 and 95 don't access dev.